src/definition/api/IApi.ts

```diff
@@ -0,0 +1,82 @@
+import { IHttp, IModify, IPersistence, IRead } from '../accessors';
+import { IApiExample } from './IApiExample';
+import { IApiEndpointInfo } from './IEndpoint';
+import { IApiRequest } from './IRequest';
+import { IApiResponse } from './IResponse';
+
+export interface IApiEndpoint {
+    /**
+     * Path to complete the api URL. Example: https://{your-server-address}/api/apps/public/{your-app-id}/{path}
+     * or https://{your-server-address}/api/apps/private/{your-app-id}/{private-hash}/{path}
+     */
+    path: string;
+    examples?: {[key: string]: IApiExample};
+
+    /**
+     * Called whenever the publically accessible url for this App is called,
+     * if you handle the methods differently then split it out so your code doesn't get too big.
+     */
+    get?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    post?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    put?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    delete?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    head?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    options?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+    patch?(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse>;
+}
+
+/**
+ * Represents an api that is being provided.
+ */
+export interface IApi {
+    /**
+     * Provides the visibility method of the URL, see the ApiVisibility descriptions for more information
+     */
+    visibility: ApiVisibility;
+    /**
+     * Provides the visibility method of the URL, see the ApiSecurity descriptions for more information
+     */
+    security: ApiSecurity;
+    /**
+     * Provide enpoints for this api registry
+     */
+    endpoints: Array<IApiEndpoint>;
+}
+
+export enum ApiVisibility {
+    /**
+     * Public api's have a fixed URL format, they are easy to remember
+     * but they are easy to be hacked as well.
```

Comment:
Can we change the `hacked` to `guessed`? Maybe reword this to be:

`A public Api has a fixed format for a url. Using it enables an easy to remember structure, however, it also means the url is intelligently guessed. As a result, we recommend having some sort of security setup if you must have a public api. Whether you use the provided security, ApiSecurity, or implement your own.`